tools/lister.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
+from __future__ import print_function
+
+import os
+import sys
+import subprocess
+import re
+from collections import defaultdict
+
+def get_ftype(fpath, use_shebang):
+    ext = os.path.splitext(fpath)[1]
+    if ext:
+        return ext[1:]
+    elif use_shebang:
+        # opening a file may throw an OSError
+        with open(fpath) as f:
+            first_line = f.readline()
+            if re.search(r'^#!.*\bpython', first_line):
+                return 'py'
+            elif re.search(r'^#!.*sh', first_line):
+                return 'sh'
+            elif re.search(r'^#!.*\bperl', first_line):
+                return 'pl'
+            elif re.search(r'^#!', first_line):
+                print('Error: Unknown shebang in file "%s":\n%s' % (fpath, first_line), file=sys.stderr)
+                return ''
+            else:
+                return ''
+    else:
+        return ''
+
+def list_files(targets=None, ftypes=None, use_shebang=True, modified_only=False,
+    exclude=None, group_by_ftype=False):
+    """
+    List all files tracked by git.
+    If targets is not None, only those files will be returned which are either
+    in targets or they are in a directory in targets.
+
+    Arguments:
+
+    targets - List of paths to directories where search is to be performed.
+        None searches current directory, [] returns [] or {}.
+    ftypes - List of file types on which to filter the search.
+        None includes all file types, [] returns [] or {}.
+    use_shebang - Determine file type of extensionless files from their shebang.
+    modified-only - Only include files which have been modified.
+    exclude - List of paths to be excluded. None acts like [].
+    group_by_ftype - If True, returns a dict of lists keyed by file type.
+        If False, returns a flat list of files.
+    """
+    if targets==[] or ftypes==[]:
+        return []
+    targets = targets or []
+    exclude = exclude or []
+    if ftypes==None:
+        ftypes_set = set()
+    else:
+        ftypes = [x.strip('.') for x in ftypes]
+        ftypes_set = set(ftypes)
+
+    cmdline = ['git', 'ls-files'] + targets
+    if modified_only:
+        cmdline.append('-m')
+
+    files_gen = (x.strip() for x in subprocess.check_output(cmdline, universal_newlines=True).split('\n'))
+    files = list(filter(os.path.isfile, files_gen)) # this also filters out empty lines
+
+    result = defaultdict(list) if group_by_ftype else []
+
+    for fpath in files:
+        # this will take a long time if exclude is very large
+        in_exclude = False
+        for expath in exclude:
+            if expath.endswith('/'):
+                expath = expath[:-1]
+            if fpath==expath or fpath.startswith(expath+'/'):
```

Comment:
use `fpath == expath` -- `==` should always have whitespace around it.  It probably makes sense to add a '\w==\w' regexp to the linter.